File: pe-math/src/main/java/org/saar/maths/transform/Rotation.java
package org.saar.maths.transform;

import org.joml.Quaternionf;
import org.joml.Quaternionfc;
import org.joml.Vector3f;
import org.joml.Vector3fc;
import org.jproperty.*;
import org.saar.maths.Angle;
import org.saar.maths.utils.Quaternion;
import org.saar.maths.utils.Vector3;

public class Rotation implements ReadonlyRotation {

    private final Quaternionf value = Quaternion.create();
    private final Vector3f eulerAngles = Vector3.create();

    private SubscribersHelper<Quaternionfc> helper = SubscribersHelper.empty();

    private Rotation(Quaternionfc value) {
        this.value.set(value);
    }

    public static Rotation fromEulerAngles(Vector3fc eulerAngles) {
        return fromEulerAngles(eulerAngles.x(), eulerAngles.y(), eulerAngles.z());
    }

    public static Rotation fromEulerAngles(float x, float y, float z) {
        return new Rotation(Quaternion.create().rotateXYZ(x, y, z));
    }

    public static Rotation fromQuaternion(Quaternionfc quaternion) {
        return new Rotation(Quaternion.of(quaternion));
    }

    public static Rotation fromQuaternion(float x, float y, float z, float w) {
        return new Rotation(Quaternion.of(x, y, z, w));
    }

    public static Rotation create() {
        return new Rotation(Quaternion.create());
    }

    private Quaternionfc copyValue() {
        return Quaternion.of(this.value);
    }

    public void set(ReadonlyRotation rotation) {
        final Quaternionfc old = copyValue();
        this.value.set(rotation.getValue());
        onChange(old);
    }

    public void set(Quaternionfc rotation) {
        final Quaternionfc old = copyValue();
        this.value.set(rotation);
        onChange(old);
    }

    public void setRotation(Angle x, Angle y, Angle z) {
        final Quaternionfc old = copyValue();
        this.value.rotateXYZ(x.getRadians(),
                y.getRadians(), z.getRadians());
        onChange(old);
    }

    public void rotate(Angle x, Angle y, Angle z) {
        rotateRadians(x.getRadians(), y.getRadians(), z.getRadians());
    }

    public void rotateDegrees(float x, float y, float z) {
        final float xRadians = (float) Math.toRadians(x);
        final float yRadians = (float) Math.toRadians(y);
        final float zRadians = (float) Math.toRadians(z);
        rotateRadians(xRadians, yRadians, zRadians);
    }

    public void rotateRadians(float x, float y, float z) {
        final Quaternionfc old = copyValue();
        this.value.rotateX(x)
                .rotateLocalY(y).rotateZ(z);
        onChange(old);
    }

    public void rotate(Rotation rotation) {
        rotate(rotation.getValue());
    }

    public void rotate(Quaternionfc rotation) {
        final Quaternionfc old = copyValue();
        this.value.mul(rotation);
        onChange(old);
    }

    private void onChange(Quaternionfc old) {
        if (!old.equals(getValue())) {
            final ChangeEvent<Quaternionfc> event =
                    new ChangeEventBase<>(this, old, getValue());
            this.helper.fireEvent(event);
        }
    }

    public void lookAlong(Vector3fc direction) {
        direction = Vector3.normalize(direction);
        if (direction.equals(Vector3.DOWN, 0)) {
            set(Quaternion.of(-1, 0, 0, 1).normalize());
        } else if (!direction.equals(Vector3.ZERO, 0)) {
            set(Quaternion.create().lookAlong(
                    direction, Vector3.UP).conjugate());
        }
    }

    @Override
    public Subscription subscribe(Subscriber<? super Quaternionfc> subscriber) {
        this.helper = this.helper.addSubscriber(subscriber);

        return () -> this.helper = this.helper.removeSubscriber(subscriber);
    }

    @Override
    public Quaternionfc getValue() {
        return this.value;
    }

    @Override
    public Vector3fc getEulerAngles() {
        final Quaternionfc value = this.value;
        return value.getEulerAnglesXYZ(this.eulerAngles);
    }

    @Override
    public Vector3f getDirection() {
        return Vector3.forward().rotate(getValue());
    }

    @Override
    public String toString() {
        return "Rotation{" + getValue() + '}';
    }
}

File: pe-math/src/main/kotlin/org/saar/maths/transform/Rotation.kt
package org.saar.maths.transform

import org.joml.*
import org.joml.Anglef.Companion.degrees
import org.jproperty.value.ObservableValueBase
import org.saar.maths.utils.Quaternion
import org.saar.maths.utils.Vector3

class Rotation private constructor(value: Quaternionfc) : ObservableValueBase<Quaternionfc>(), ReadonlyRotation {

    private val value = Quaternion.of(value)

    override val eulerAngles: Vector3f = Vector3.create()
        get() {
            val value = this.value
            return value.getEulerAnglesXYZ(field)
        }

    private fun copyValue(): Quaternionfc {
        return Quaternion.of(this.value)
    }

    fun set(rotation: ReadonlyRotation) {
        val old = copyValue()
        this.value.set(rotation.value)
        onChange(old)
    }

    fun set(rotation: Quaternionfc) {
        val old = copyValue()
        this.value.set(rotation)
        onChange(old)
    }

    @JvmSynthetic
    fun rotate(x: Anglef, y: Anglef, z: Anglef) {
        val old = copyValue()
        this.value.rotateX(x)
            .rotateLocalY(y)
            .rotateZ(z)
        onChange(old)
    }

    fun rotateDegrees(x: Float, y: Float, z: Float) = rotate(x.degrees, y.degrees, z.degrees)

    fun rotate(rotation: Rotation) {
        rotate(rotation.value)
    }

    fun rotate(rotation: Quaternionfc) {
        val old = copyValue()
        this.value.mul(rotation)
        onChange(old)
    }

    private fun onChange(old: Quaternionfc) {
        if (old != this.value) {
            fireChangeEvent(old)
        }
    }

    fun lookAlong(direction: Vector3fc) {
        var direction = direction
        direction = Vector3.normalize(direction)
        if (direction.equals(Vector3.DOWN, 0f)) {
            set(Quaternion.of(-1f, 0f, 0f, 1f).normalize())
        } else if (!direction.equals(Vector3.ZERO, 0f)) {
            set(
                Quaternion.create().lookAlong(
                    direction, Vector3.UP
                ).conjugate()
            )
        }
    }

    override fun getValue(): Quaternionfc {
        return this.value
    }

    override val direction: Vector3f
        get() = Vector3.forward().rotate(this.value)

    override fun toString(): String {
        return "Rotation{" + this.value + '}'
    }

    companion object {
        fun fromEulerAngles(eulerAngles: Vector3fc): Rotation {
            return fromEulerAngles(eulerAngles.x(), eulerAngles.y(), eulerAngles.z())
        }

        fun fromEulerAngles(x: Float, y: Float, z: Float): Rotation {
            return Rotation(Quaternion.create().rotateXYZ(x, y, z))
        }

        fun fromQuaternion(quaternion: Quaternionfc): Rotation {
            return Rotation(Quaternion.of(quaternion))
        }

        fun fromQuaternion(x: Float, y: Float, z: Float, w: Float): Rotation {
            return Rotation(Quaternion.of(x, y, z, w))
        }

        fun create(): Rotation {
            return Rotation(Quaternion.create())
        }
    }
}
